### PR TITLE
Restrict mappers from configured role

### LINKED ISSUE #5584: [RFC] Allow admin mappers to restrict their contents by role
## Feature Request
Allow the following methods to restrict their contents based on a new "role" option passed in their 3rd argument:
* `ListMapper::add()`
* `DatagridMapper::add()`
* `ShowMapper::add()`
* `ShowMapper::with()`
* `ShowMapper::tab()`
* `FormMapper::add()`
* `FormMapper::with()`
* `FormMapper::tab()`

I think the easiest approach to achieve this behavior is to extend this condition
https://github.com/sonata-project/SonataAdminBundle/blob/64b83e290566dc609a7a30d6a419c1359430dfc9/src/Form/FormMapper.php#L59
with the following check
```php
if (null !== $this->apply && !$this->apply || (isset($fieldDescriptionOptions['role']) && !$this->admin->isGranted($fieldDescriptionOptions['role']))) {
```

So, if the authenticated user doesn't have the required role, the call to any of these methods will be ignored:
```php
 $showMapper->add('someProp', null, ['role' => 'ROLE_VIEW_SOME_PROP']);
```

### PR DESCRIPTION
**Pull request by phansys**
## Subject

Restrict mappers from configured role

I am targeting this branch, because these changes are fully BC.

Closes #5584.

## Changelog

```markdown
### Added
- Added "role" option for methods `ListMapper::add()`, `DatagridMapper::add()`, `ShowMapper::add()` and `FormMapper::add()`; in order to restrict the content rendering based on the provided role.
```
    
## To do
    
- [x] Update remaining methods;
- [x] Add tests;
- [x] Update the documentation.

### CODE REVIEW

**Comment by phansys** [comment]
Sorry, the **TO DO** list was hidden by comment delimiters.

**Comment by phansys** [comment]
I have a concern regarding the behavior of this feature. Currently, I'm trying to not return early if the required role is not granted, allowing the method's logic to perform any further validation like sanity checks, options resolving and so.
Watching the existing codebase, I saw the behavior of `ifTrue()` and `ifFalse()` methods, which perform a similar task, but just for fields (it doesn't avoid the creation of a group or a tab, just the creation of their child fields if used at a higher level):
https://github.com/sonata-project/SonataAdminBundle/blob/64b83e290566dc609a7a30d6a419c1359430dfc9/src/Form/FormMapper.php#L59-L61
This feature isn't allowing any of the previously mentioned checks, so my question is: should the role based restriction follow the same premise or should I keep it as currently is?

**Comment by phansys** [comment]
By the moment I have simplified the first approach, leaving out of this feature the grouping methods (`ShowMapper::with()`, `ShowMapper::tab()`, `FormMapper::with()` and `FormMapper::tab()`). Depending on the required effort and the feedback, they could be taken into account later.

File: src/Form/FormMapper.php [hunk 1]
```diff
@@ -93,7 +93,7 @@ public function add($name, $type = null, array $options = [], array $fieldDescri
             $fieldDescriptionOptions['type'] = $type;
         }
 
-        if ($group['translation_domain'] && !isset($fieldDescriptionOptions['translation_domain'])) {
+        if (!isset($fieldDescriptionOptions['translation_domain']) && isset($group['translation_domain'])) {
```
Comment (greg0ire) [inline src/Form/FormMapper.php]: Why this change? Are you also fixing a bug?

Comment (phansys) [reply src/Form/FormMapper.php]: No, it was a change I made while covering `with()` and `tab()` methods. It's reverted now, thank you!

File: src/Form/FormMapper.php [hunk 2]
```diff
@@ -114,9 +114,13 @@ public function add($name, $type = null, array $options = [], array $fieldDescri
 
         if ($name instanceof FormBuilderInterface) {
             $this->formBuilder->add($name);
+            $type = null;
```
Comment (kunicmarko20) [inline src/Form/FormMapper.php]: why is this needed here?

Comment (kunicmarko20) [inline src/Form/FormMapper.php]: why is this needed here?

Comment (kunicmarko20) [inline src/Form/FormMapper.php]: why is this needed here?

File: src/Form/FormMapper.php [hunk 3]
```diff
@@ -134,13 +138,15 @@ public function add($name, $type = null, array $options = [], array $fieldDescri
...
-            $this->formBuilder->add($fieldDescription->getName(), $type, $options);
-
             if (null !== $help) {
-                $this->admin->getFormFieldDescription($fieldDescription->getName())->setHelp($help);
+                $this->admin->getFormFieldDescription($name)->setHelp($help);
             }
         }
 
+        if (!isset($fieldDescriptionOptions['role']) || $this->admin->isGranted($fieldDescriptionOptions['role'])) {
+            $this->formBuilder->add($name, $type, $options);
```
Comment (kunicmarko20) [inline src/Form/FormMapper.php]: line #116 already adds it, won't this add it again?

File: src/Form/FormMapper.php (hunk 2 above)
Comment (phansys) [reply src/Form/FormMapper.php]: Because if the previous condition is met, we need to pass default values to arguments 2 and 3 for `FormBuilder::add()`.

File: src/Form/FormMapper.php (hunk 3 above)
Comment (phansys) [reply src/Form/FormMapper.php]: Yes, that call was removed. Thank you!

File: docs/reference/architecture.rst [hunk 4]
```diff
@@ -196,9 +196,21 @@ which stores instances of ``FieldDescriptionInterface``. Picking up on our previ
...
+                // user is granted with the "ROLE_ADMIN_MODERATOR" role
+                ->add('privateNotes', null, [], [
+                    'role' => 'ROLE_ADMIN_MODERATOR'
+                ])
+
                 // if no type is specified, SonataAdminBundle tries to guess it
                 ->add('body')
 
+                // conditionally add "status" field if the subject already exists
+                // `ifFalse()` is also available to build this kind of conditions
```
Comment (greg0ire) [inline docs/reference/architecture.rst]:
```suggestion
                // `ifFalse()` is also available to build this kind of condition
```

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

**Comment by phansys** [comment]
Commits are squashed :+1:

**Comment by greg0ire** [comment]
Thanks @phansys !
